Implement torch.broadcast_tensors

zou3519 opened the pull request:
This exposes expand_outplace to python. Fixes #8076. Fixes #10041.

I didn't name it torch.broadcast because numpy.broadcast does something
slightly different (it returns an object with the correct shape
information).

Test Plan: new test_torch, test_autograd tests.

cc @soumith @fritzo

zou3519 committed:
Implement torch.broadcast_all

This exposes expand_outplace to python. Fixes #8076. Fixes #10041.

I didn't name it torch.broadcast because numpy.broadcast does something
slightly different (it returns an object with the correct shape
information).

Test Plan: new test_torch, test_autograd tests.

fmassa commented:
Nice!
I believe the equivalent [numpy function is `numpy.broadcast_arrays`](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.broadcast_arrays.html#numpy.broadcast_arrays), should we aim at maintaining the same name?

soumith commented:
yes let's match numpy and do `broadcast_arrays`

fritzo commented:
Nice! The additional behavior of `torch.distributions.utils.broadcast_all` is to handle python floats. I believe we can update `broadcast_all` to use your `broadcast_arrays` as follows (in either this PR or a follow-up PR):
```py
# in torch/distributions/utils.py
def broadcast_all(*values):
    """docstring"""
    if not all(map(torch.is_tensor, values)):
        # promote floats to tensors
        new_tensor = torch.tensor
        for value in values:
            if torch.is_tensor(value):
                new_tensor = value.new_tensor
                break
        values = [v if torch.is_tensor(v) else new_tensor(v) for v in values]
    return torch.broadcast_arrays(*values)
```
This would also be a great test to see that things still work 😄

vadimkantorov commented:
Or even put the scalar broadcasting in `torch.broadcast_all/arrays`? :) If `torch.where` uses it, it would solve the `torch.where` not accepting scalars.

fmassa commented:
@vadimkantorov I think there was a discussion around using `torch.as_tensor` (or something equivalent) in the beginning of every PyTorch function, so that it also works for numbers and numpy arrays.
Not sure what's the status of that. @gchanan did we decide on something about that?

zou3519 committed:
Some changes:

- s/broadcast_all/broadcast_tensors/
- broadcast_tensors now takes varargs

zou3519 committed:
Reimplement broadcast_all in torch/distributions/utils.

zou3519 committed:
Python 2 compatibility

zou3519 commented:
This should be good for review, despite the hanging tests.

I updated the following:
- renamed torch.broadcast_all to torch.broadcast_tensors (@fmassa)
- Changed torch.distributions.utils.broadcast_all() to use `torch.broadcast_tensors` (@fritzo). All the distribution tests pass so it looks like it's good to go :)

fritzo commented on torch/distributions/utils.py:
```diff
@@ -64,22 +48,20 @@ def broadcast_all(*values):
-    return values
+    if not all(map(torch.is_tensor, values)):
+        # promote numbers to tensors of dtype torch.get_default_dtype()
+        def default_promotion(v):
```
nit: We should define this at file level to avoid the overhead of creating a new function on each invocation.
```py
def _default_promotion(v):
    return torch.tensor(v, dtype=torch.get_default_dtype())
```

fritzo commented on torch/distributions/utils.py:
```diff
@@ -64,22 +48,20 @@ def broadcast_all(*values):
-    scalar_idxs = [i for i in range(len(values)) if isinstance(values[i], Number)]
-    tensor_idxs = [i for i in range(len(values)) if values[i].__class__.__name__ == 'Tensor']
-    if len(scalar_idxs) + len(tensor_idxs) != len(values):
+    if not all(map(lambda v: torch.is_tensor(v) or isinstance(v, Number), values)):
```
nit: it's a tad bit more readable to 
```py
if not all(torch.is_tensor(v) or isinstance(v, Number) for v in values):
    ...
```

zou3519 committed:
Address nits

fritzo commented:
This change is great. How can I start to use it? It does not appear to have landed in master yet.

zou3519 commented:
It should be on master, the following works for me on a latest checkout:
```
import torch
x = torch.randn(2, 2)
y = torch.randn(1)
torch.broadcast_tensors(x, y)
```
